locked-asset/simple-lock/src/proxy_farm.rs

```diff
@@ -0,0 +1,303 @@
+elrond_wasm::imports!();
+elrond_wasm::derive_imports!();
+
+use crate::{error_messages::*, locked_token::LockedTokenAttributes};
+
+#[derive(
+    TypeAbi, TopEncode, TopDecode, NestedEncode, NestedDecode, PartialEq, Debug, Clone, Copy,
+)]
+pub enum FarmType {
+    SimpleFarm,
+    FarmWithLockedRewards,
+}
+
+#[derive(TypeAbi, TopEncode, TopDecode, NestedEncode, NestedDecode, PartialEq, Debug)]
+pub struct FarmProxyTokenAttributes<M: ManagedTypeApi> {
+    pub farm_type: FarmType,
+    pub farm_token_id: TokenIdentifier<M>,
+    pub farm_token_nonce: u64,
+    pub farming_token_id: TokenIdentifier<M>,
+    pub farming_token_locked_nonce: u64,
+}
+
+pub type EnterFarmThroughProxyResultType<M> = EsdtTokenPayment<M>;
+pub type ExitFarmThroughProxyResultType<M> = MultiValue2<EsdtTokenPayment<M>, EsdtTokenPayment<M>>;
+pub type FarmClaimRewardsThroughProxyResultType<M> =
+    MultiValue2<EsdtTokenPayment<M>, EsdtTokenPayment<M>>;
+pub type FarmCompoundRewardsThroughProxyResultType<M> = EsdtTokenPayment<M>;
+
+#[elrond_wasm::module]
+pub trait ProxyFarmModule:
+    crate::farm_interactions::FarmInteractionsModule
+    + crate::locked_token::LockedTokenModule
+    + elrond_wasm_modules::default_issue_callbacks::DefaultIssueCallbacksModule
+{
+    #[only_owner]
+    #[payable("EGLD")]
+    #[endpoint(issueFarmProxyToken)]
```

Comment:
I think we can have another strategy for these kinds of setups - as we do not necessary need to do this from a contract. We could issue the token, setroles for the contract and give the ownership to the contract. so we will not need to do these proxy stuff - in a set of contract this might work @catalinnnn @claudiu725 ?

Reply:
It could work, but I think the current implementation needs fewer setup steps, and also ensures that everything worked as intended (as long as the transactions went through successfully)

In the current version, we only need two transactions: issue and set roles.
In the version you're suggesting, we would need more steps:
- issue the token
- get the token ID from the issue transaction
- set the local roles on the contract's address
- change token ownership to the contract
- set the token ID in the contract

Reply:
I think manual issue and setting roles are the way to go, especially if we have the ability to transfer the token ownership to the contract, which I didn't know about

Reply:
You can transfer ownership of tokenID to a contract.